sandbox/apparmor/notify/message.go

```diff
@@ -0,0 +1,418 @@
+
+	// Unpack the base structure.
+	if err := msg.MsgHeader.UnmarshalBinary(data); err != nil {
+		return err
+	}
+
+	// Unpack fixed-size elements.
+	buf := bytes.NewBuffer(data)
+	var raw msgNotificationFilterKernel
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Read(buf, order, &raw); err != nil {
+		return fmt.Errorf("%s: cannot unpack: %s", prefix, err)
+	}
+
+	// Unpack variable length elements.
+	unpacker := newStringUnpacker(data)
+	ns, err := unpacker.UnpackString(raw.NS)
+	if err != nil {
+		return fmt.Errorf("%s: cannot unpack namespace: %v", prefix, err)
+	}
+	filter, err := unpacker.UnpackString(raw.Filter)
+	if err != nil {
+		return fmt.Errorf("%s: cannot unpack filter: %v", prefix, err)
+	}
+
+	// Put everything together.
+	msg.ModeSet = ModeSet(raw.ModeSet)
+	msg.NameSpace = ns
+	msg.Filter = filter
+
+	return nil
+}
+
+// MarshalBinary marshals the message into binary form.
+func (msg *MsgNotificationFilter) MarshalBinary() (data []byte, err error) {
+	var raw msgNotificationFilterKernel
+	packer := newStringPacker(raw)
+	raw.Version = 2
+	raw.ModeSet = uint32(msg.ModeSet)
+	raw.NS = packer.PackString(msg.NameSpace)
+	raw.Filter = packer.PackString(msg.Filter)
+	raw.Length = packer.TotalLen()
+	msgBuf := bytes.NewBuffer(make([]byte, 0, raw.Length))
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Write(msgBuf, order, raw); err != nil {
+		return nil, err
+	}
+	if _, err := msgBuf.Write(packer.Bytes()); err != nil {
+		return nil, err
+	}
+	return msgBuf.Bytes(), nil
+}
+
+// Validate returns an error if the mssage contains invalid data.
+func (msg *MsgNotificationFilter) Validate() error {
+	if !msg.ModeSet.IsValid() {
+		return fmt.Errorf("unsupported modeset: %d", msg.ModeSet)
+	}
+	return nil
+}
+
+// MsgNotification describes a kernel notification message.
+//
+// This structure corresponds to the kernel type struct apparmor_notif
+// described below.
+//
+//	struct apparmor_notif {
+//	  struct apparmor_notif_common base;
+//	  __u16 ntype;        /* notify type */
+//	  __u8 signalled;
+//	  __u8 flags;
+//	  __u64 id;           /* unique id, not globally unique*/
+//	  __s32 error;        /* error if unchanged */
+//	} __attribute__((packed));
+type MsgNotification struct {
+	MsgHeader
+	// NotificationType describes the kind of notification message used.
+	// Currently the kernel only sends APPARMOR_NOTIF_OP messages.
+	NotificationType NotificationType
+	// XXX: Signaled seems to be unused.
+	Signalled uint8
+	// Set to 1 to NOT cache
+	Flags uint8
+	// ID is an opaque kernel identifier of the notification message. It must be
+	// repeated in the MsgNotificationResponse if one is sent back.
+	ID uint64
+	// XXX: This seems to be unused and clashes with identical field in
+	// MsgNotificationOp.
```

Comment:
Is there an error in both the `struct apparmor_notif_common base` field and the `error` field of `struct apparmor_notif`, and should either/both be set to anything besides `0`?

Reply:
From John's comment below: `Error` here is the error which the kernel returns to the application if the notification is denied.